source/rust_verify/src/verifier.rs

```diff
@@ -231,6 +236,67 @@ impl Verifier {
         }
     }
 
+    fn print_profile_stats(
+        &self,
+        compiler: &Compiler,
+        profiler: Profiler,
+        qid_map: &HashMap<String, vir::sst::Exp>,
+    ) {
+        let num_quants = profiler.quant_count();
+        let total = profiler.total_instantiations();
+        let max = 10;
+        let delimiter = "-".repeat(100);
+        let msg = format!(
+            "Observed {} total instantiations of user-level quantifiers",
+            total.to_formatted_string(&Locale::en)
+        );
+        compiler.diagnostic().note_without_error(&msg);
+
+        for (index, cost) in profiler.iter().take(max).enumerate() {
+            println!("{}", delimiter);
+            // Report the quantifier
+            let qexp = qid_map
+                .get(&cost.quant)
+                .expect(format!("Failed to find quantifier {}", cost.quant).as_str());
+            let span = from_raw_span(&qexp.span.raw_span);
+            let count = cost.instantiations;
+            let msg = format!(
+                "Cost * Instantiations: {} (Instantiated {} times - {}% of the total, cost {}) top {} of {} user-level quantifiers.\n",
+                count * cost.cost,
+                count.to_formatted_string(&Locale::en),
+                100 * count / total,
+                cost.cost,
+                index + 1,
+                num_quants
+            );
+            compiler.diagnostic().span_note_without_error(span, &msg);
+
+            let triggers = match &qexp.x {
```

Comment:
I was a little uncomfortable that you're carrying a reference to the `Exp` in the `qid_map`, but I can see why.
What would you think about only stashing the `Span` and the trigs in the map, i.e. moving this logic to the point where you add quants to the map? (The span is tiny and the triggers are already `Arc`ed)

Reply:
Done.